refactor(options): remove `set_string_option_direct()`

famiu opened the pull request:
Problem: `set_string_option_direct()` contains a separate codepath specifically for setting string options. Not only is that unnecessary code duplication, but it's also limited to only string options.

Solution: Replace `set_string_option_direct()` with `set_option_direct()` which calls `set_option()` under the hood. This reduces code duplication and allows directly setting an option of any type.

Supersedes #25855.

famiu committed:
refactor(options): restructure `set_option`

zeertzjq commented on src/nvim/option.c:
```diff
@@ -3601,91 +3632,105 @@ static const char *did_set_option(OptIndex opt_idx, void *varp, OptVal old_value
+  // For local scope of a global-local option, it's equal to the global value if the local value is
+  // unset. In every other case, it is the same as old_value.
+  // This value is used instead of old_value when triggering the OptionSet autocommand.
+  OptVal used_old_value = (scope_local && is_opt_local_unset)
+                          ? optval_from_varp(opt_idx, get_varp(opt))
+                          : old_value;
```
Hmm, this changes the behavior of `OptionSet`.

famiu committed:
refactor(options): remove `set_string_option_direct()`

Problem: `set_string_option_direct()` contains a separate codepath specifically for setting string options. Not only is that unnecessary code duplication, but it's also limited to only string options.

Solution: Replace `set_string_option_direct()` with `set_option_direct()` which calls `set_option()` under the hood. This reduces code duplication and allows directly setting an option of any type.

famiu committed:
refactor(options): add `direct` flag to `set_option()`

justinmk commented:
# WIP: debug failing test

> The tests that are failing are in swapfile_preserve_recover_spec.lua

glacambre commented:
@famiu this should fix the problem:
```diff
diff --git a/src/nvim/option.c b/src/nvim/option.c
index d7c46a44e..fe3209844 100644
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -3459,12 +3459,12 @@ static bool is_option_local_value_unset(OptIndex opt_idx)
 /// @param       errbuflen       Length of error buffer.
 ///
 /// @return  NULL on success, an untranslated error message on error.
-static const char *did_set_option(OptIndex opt_idx, void *varp, OptVal old_value, int opt_flags,
-                                  scid_T set_sid, const bool direct, const bool value_replaced,
+static const char *did_set_option(OptIndex opt_idx, void *varp, OptVal new_value, OptVal old_value,
+                                  int opt_flags, scid_T set_sid, const bool direct,
+                                  const bool value_replaced,
                                   char *errbuf,  // NOLINT(readability-non-const-parameter)
                                   size_t errbuflen)
 {
-  OptVal new_value = optval_from_varp(opt_idx, varp);
   vimoption_T *opt = &options[opt_idx];
   const char *errmsg = NULL;
   bool restore_chartab = false;
@@ -3741,8 +3741,8 @@ static const char *set_option(const OptIndex opt_idx, void *varp, OptVal value,
   // Set option through its variable pointer.
   set_option_varp(opt_idx, varp, value, false);
   // Process any side effects.
-  errmsg = did_set_option(opt_idx, varp, old_value, opt_flags, set_sid, direct, value_replaced,
-                          errbuf, errbuflen);
+  errmsg = did_set_option(opt_idx, varp, value, old_value, opt_flags, set_sid, direct,
+                          value_replaced, errbuf, errbuflen);
 
   secure = secure_saved;
```
The problem was that you used `optval_from_varp(opt_idx, varp)` to get the OptVal to use in `did_set_option`, but this would end up synthetizing a new value for `&modified` instead of using the one set by the user.

I would also recommend adding a new test that checks for `:set fileformat=dos nomodied` and `:set fileformat=unix nomodified` and splitting the `set` calls in the currently failing test in order to make debugging these tests easier in the future.

famiu committed:
test(options): add test for `'modified'`

justinmk commented:
coverity:

```
*** CID 488299:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/option.c: 3646 in validate_option_value()
3640           optval_free(*newval);
3641           *newval = optval_copy(get_option_unset_value(opt_idx));
3642         }
3643       } else if (!option_has_type(opt_idx, newval->type)) {
3644         char *rep = optval_to_cstr(*newval);
3645         char *valid_types = option_get_valid_types(opt_idx);
>>>     CID 488299:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "rep" to "snprintf", which expects a null-terminated string.
3646         snprintf(errbuf, IOSIZE, _("Invalid value for option '%s': expected %s, got %s %s"),
3647                  opt->fullname, valid_types, optval_type_get_name(newval->type), rep);
3648         xfree(rep);
3649         xfree(valid_types);
3650         errmsg = errbuf;
3651       } else if (newval->type == kOptValTypeNumber) {
```
